### ISSUE TITLE
Offline Collector GCS does not have delete access

### ISSUE THREAD

**Issue by Zawadidone** [Velocidex/velociraptor#2542]
1. GCS bucket is created 
2. Service account with the GCS role `roles/storage.objectCreator` is created.    https://cloud.google.com/storage/docs/access-control/iam-roles
3. Offline collector is created and executed on a system
4. The offline collector is created successfully and uploaded to the GCS Bucket. But the log file can not be uploaded after the collection is done, because at the start the same file name was created in the bucket. And the service account does not have delete permission on the bucket.

```
[INFO] TIMESTAMP Uploading to COLLECTION.log
[INFO] TIMESTAMP upload_gcs: Uploading COLLECTION.log to GCS_BUCKET
[INFO] TIMESTAMP upload_gcs COLLECTION.log: Uploaded 45 bytes
[INFO] TIMESTAMP upload_gcs: SUCCESS writing to object: [...]
[INFO] TIMESTAMP upload_gcs: GCS Calculated MD5 [...]
[...] - Collection of files 
[INFO] TIMESTAMP upload_gcs: ERROR writing to object: googleapi: Error 403: SERVICE_ACCOUNT does not have storage.objects.delete access to the Google Cloud Storage object., forbidden
```

I am aware that the blog post (https://velociraptor.velocidex.com/triage-with-velociraptor-pt-3-d6f63215f579) uses a service account that has the delete permission for objects `roles/storage.legacyBucketWriter`. I prefer not to use this permission in the case an unauthorized user has access to the private key of the service account in the offline collector. I have not taken in accounted if this is also the case for the other collection types like AWS.

Is it possible that the log file is uploaded after the offline collector is done and not at the start?

**Comment by scudette** [comment]
We need to upload something before we start in order to establish that we have the right credentials to upload before we start to do anything. It would be very annoying to go through the whole acquisition phase and then find our creds are not good to actually upload the file!

I think adding delete permission to the service account is fine as long as they dont have list permissions. The filename we use is random enough to be hard to guess so it is unlikely that the key can be used to delete existing files.

Finally in recent versions of Velociraptor it is not really necessary to get the log file anyway since it is in the zip file mostly (except for the final hash of course which can only be finalized after the zip is closed). 

Maybe we need to just write a hash file separately ?

**Comment by Zawadidone** [comment]
@scudette do you have a suggestion for a fix to which I could possible contribute?

**Comment by scudette** [comment]
I think this is fixed in 0.7.0 where we don't actually stop the collection if we fail to upload the file, instead we leave the zip file behind to let the user sort out another way to send it

**Comment by Zawadidone** [comment]
Okay I have used the branches `v0.7.0` and `master`, created a collector for Windows to upload to a GCS bucket and I have a few questions about the changes.

**Log error**

The logs shows the following: 
```
[ERROR] 2023-08-02T01:21:10-07:00  upload_gcs: GCS Calculated MD5: e59134274f7395d94d7303e5262788ba Hash checks out.
[...]
[ERROR] 2023-08-02T01:21:10-07:00  upload_gcs: GCS Calculated MD5: 8c8de18b63cb10c4748765d54576ebdf Hash checks out.
```

That indicates that the hash checked out, but the message is logged as an error. Is this intended? 

**Collection complete**

As shown in the screenshot, the collection is successfully completed but the ZIP file and log file are left behind. Is this intended? 

I would have expected that Velociraptor would delete both files because the collection is successfully completed.


![image](https://github.com/Velocidex/velociraptor/assets/21525407/66b2c02e-6187-4255-a94f-662413611ad8)

**Comment by scudette** [comment]
Yes we now always leave the files behind because it's really hard to figure out if the upload actually worked. There were some cases with AWS that returned some results but the upload failed for some reason

I thought it's best to let the user figure it out after collecting a large collection and manually remove the files.

What do you think? 

I know this might be cumbersome when just pushing the collector with edr so maybe we need an option?

**Comment by Zawadidone** [comment]
Yes it would be nice to have an option in the Create Collector menu to delete the ZIP and log file after the collection is completed, whether it is successful or not.

**Comment by Zawadidone** [comment]
@scudette will the fix of this issue be part of version 0.7.0?

**Comment by scudette** [comment]
In 0.7.0 we no longer delete the collected files anyway because we can be 100% sure the upload succeeded so we leave it to the user to remove the files once they confirm

So now we don't really need to overwrite any log files so I expect this issue will be fixed. You can test it with the 0.7.0 RC

**Comment by Zawadidone** [comment]
The issue is fixed. I will create another issue for the option to delete the ZIP file after the collection is finished.
